### PR TITLE
Implement MetadataClient

### PR DESCRIPTION
**Pull request by fjogeleit**
Signed-off-by: Frank Jogeleit <frank.jogeleit@web.de>

### CODE REVIEW

File: pkg/kubernetes/queue.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	debouncer Debouncer
+}
+
+func (q *Queue) Add(event report.Event, obj *v1.PartialObjectMetadata) error {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return err
+	}
+
+	q.queue.Add(event.String() + ":" + key)
```
Comment (eddycharly) [inline pkg/kubernetes/queue.go]: Not sure about that, aren't you going to loose advantages of the queue here ?

Comment (fjogeleit) [reply pkg/kubernetes/queue.go]: Which advantage do you mean? After each informer event it adds the key and related event type to the queue und process it in the worker.

Comment (eddycharly) [reply pkg/kubernetes/queue.go]: The queue guarantees that for a given item in the queue, only one worker will process it at a time.
This means that if you receive multiple events, they get pushed to the queue but if a worker is already processing the item, it is requeued and processed again later.
For this to work the key has to be stable (all events triggered by the same item need to produce the same key).

Comment (fjogeleit) [reply pkg/kubernetes/queue.go]: Ah okay, understand. So I need some kind of cache to get information about the type of underlying informer event

Comment (eddycharly) [reply pkg/kubernetes/queue.go]: Why ? Do you need to know the underlying event type ?
When you call `Get` if the object is not found, you can consider it a deletion. If it is found, it's either a creation or an update, if you don't have it in your own cache you can consider it a creation, if you have it in your own cache you can consider it's an update.

You already have your own cache right ?

Comment (fjogeleit) [reply pkg/kubernetes/queue.go]: hm yes, thats right. I will try it out. Thanks a lot as always.

File: pkg/config/resolver.go [hunk 2]
```diff
@@ -303,6 +335,16 @@ func (r *Resolver) ReportFilter() *report.Filter {
...
+	}
+
+	r.cache = cache.New()
+
+	return r.cache
+}
+
 // ResultCache resolver method
 func (r *Resolver) ResultCache() cache.Cache {
 	if r.resultCache != nil {
```
[the remark this answers is not in the document]
Comment (fjogeleit) [reply pkg/config/resolver.go:349]: @sonatype-lift ignore

File: pkg/kubernetes/fixtures_test.go [hunk 3]
```diff
@@ -5,10 +5,21 @@ import (
...
 	"github.com/kyverno/policy-reporter/pkg/report"
 
+	pr "github.com/kyverno/policy-reporter/pkg/crd/api/policyreport/v1alpha2"
 	"github.com/kyverno/policy-reporter/pkg/crd/client/clientset/versioned/fake"
 	v1alpha2client "github.com/kyverno/policy-reporter/pkg/crd/client/clientset/versioned/typed/policyreport/v1alpha2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	metafake "k8s.io/client-go/metadata/fake"
 )
 
+func NewFakeMetaCilent() (*metafake.FakeMetadataClient, metafake.MetadataClient, metafake.MetadataClient) {
```
Comment (eddycharly) [inline pkg/kubernetes/fixtures_test.go]: Typo :)

Comment (fjogeleit) [reply pkg/kubernetes/fixtures_test.go]: fixed :->

File: pkg/kubernetes/fixtures_test.go [hunk 4]
```diff
@@ -5,10 +5,21 @@ import (
...
 
+func NewFakeMetaClient() (*metafake.FakeMetadataClient, metafake.MetadataClient, metafake.MetadataClient) {
+	schema := metafake.NewTestScheme()
+	metav1.AddMetaToScheme(schema)
+
+	client := metafake.NewSimpleMetadataClient(schema)
+	return client, client.Resource(pr.SchemeGroupVersion.WithResource("policyreports")).Namespace("test").(metafake.MetadataClient), client.Resource(pr.SchemeGroupVersion.WithResource("clusterpolicyreports")).(metafake.MetadataClient)
+}
+
 func NewFakeCilent() (*fake.Clientset, v1alpha2client.PolicyReportInterface, v1alpha2client.ClusterPolicyReportInterface) {
```
Comment (eddycharly) [inline pkg/kubernetes/fixtures_test.go]: Another one

**Pull request merged**